### DISCUSSION TITLE
Updating query parameters in a link

### DISCUSSION THREAD

**Discussion by adam820** [elimintz/justpy#217] [Q&A] [answered]
Howdy! Is there a built-in way to build an <a> href to a new URL that maintains the existing query params while updating one or more of them? Or would I need to build that myself out of the existing `request` object?

For example, I have a URL `http://localhost:8000/?foo=bar&foo2=baz`. I want to create a button/link with an HREF that points at the same URL, but with `buz` as the `foo2` param: `http://localhost:8000/?foo=bar&foo2=buz`. This can be done manually (or with a helper function), but I wanted to make sure I wasn't overlooking a built-in function (a la Flask's [`url_for`](https://flask.palletsprojects.com/en/1.1.x/api/#flask.url_for)).

Thanks!

**Comment by elimintz** [comment, marked as answer]
@adam820  Is this what you are looking to do:

```python
import justpy as jp

def link_test(request):
    wp = jp.WebPage()
    foo = request.query_params.get('foo', 'No foo' )
    foo2 = request.query_params.get('foo2', 'No foo2' )
    jp.Div(text=f'foo: {foo}, foo2: {foo2}', classes="m-4 text-xl", a=wp)
    jp.A(text='New link', href='http://localhost:8000/?foo=bar&foo2=buz', a=wp,
         classes='m-4 p-2 text-xl text-white bg-blue-500 hover:bg-blue-700')
    return wp

jp.justpy(link_test)
```

If not, please let me know

**Comment by adam820** [comment]
Sort of, but not quite. I figured it could be done manually, but I was checking to see if there was something that would auto-build the href with the query parameters already intact (by passing them in), to avoid having to manually hardcode addresses, etc.

I ended up making my own quick and dirty params builder function that updates an existing parameter, or otherwise accepts new parameters passed in as kwargs.

```python
from urllib.parse import urlencode

def build_url_params(req_obj, **kwargs):
    # Copy the original query parameters over to a new mutable dict
    qparams = dict()
    for k, v in req_obj.query_params.items():
        qparams[k] = v
    try:
        # Build the query params
        for k, v in kwargs.items():
            if k in qparams.keys():
                qparams[k] = v
            else:
                qparams[k] = v
    except (ValueError, KeyError):
        return "?"

    return f"?{urlencode(qparams)}"
```

**Reply by matthoendorf** [reply]
I've created click functions for buttons that use wp.redirect

```
def button_click(self, msg):
    wp = msg.page
    wp.redirect = f'/page?param1={self.param1}&param2={self.param2}'
```

Not sure if this helps...

**Reply by adam820** [reply]
Not quite, since it looks like you're hardcoding parameters in the redirect. The part of `url_for` (from Flask) that I was inquiring about is thus: _Variable arguments that are unknown to the target endpoint are appended to the generated URL as query arguments._ You can pass in an arbitrary number of keyword arguments and the URL is auto-built with whatever is already there _plus_ the new parameters.

Either way, I have moved on from this using this project at this time in my application; thanks for including that information for future readers!
